### DISCUSSION TITLE
Know in the batch-finished-callback which exact jobs in the batch failed

### DISCUSSION THREAD

**Discussion by amo13** [bensheldon/good_job#1187] [Q&A] [answered]
I have a batch of jobs running as a batch with an on-finished callback-job. I'd like this callback-job to inform the admin about how many and which exact jobs in the batch failed.
I understand that I can use the on_discard callback, but this would trigger on the first discarded job and would not know about other discarded ones.

I have tried to access some attribute of the job instances in `bulk.active_jobs`, but I don't know how to find out if any job in that array was successful or discarded. Intuitively, I'd do something like this:

```ruby
class MyUpdateCallbackJob < ApplicationJob
  queue_as :updates

  def perform(batch, params)
    discarded_jobs = bulk.active_jobs.select{|j| !j.success}

    tell_about(discarded_jobs)
  end
end
```

I also tried to add an instance variable `success` to the jobs that are enqueued as part of the batch (all same class), but I was not able to make that instance variable accessible like above from inside the callback-job.

I'd greatly appreciate any help pointing me in the right direction for this.

**Comment by bensheldon** [comment, marked as answer]
There isn't a public public api for job records, but it's simple enough to grab the Active Record models directly:

```ruby
jobs = GoodJob::Job.discarded.where(batch_id: batch.id)
```

I wish that Active Job itself had a 1st party interface for presenting job status, but it does not.

**Reply by amo13** [reply]
Thank you once more.
You should definitely be awarded the friendliest and most helpful guy there is.
